CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_ActivityChainSpawner.uc

```diff
@@ -0,0 +1,451 @@
+//---------------------------------------------------------------------------------------
+//  AUTHOR:  Xymanek
+//  PURPOSE: Replacement for base game MissionCalendar which instead spawns CAs based on
+///          "work" done by XCom (base) and contacted regions and relays built (bonus)
+//---------------------------------------------------------------------------------------
+//  WOTCStrategyOverhaul Team
+//---------------------------------------------------------------------------------------
+
+class XComGameState_ActivityChainSpawner extends XComGameState_BaseObject config(Infiltration);
+
+var protectedwrite float PreviousWork;
+var protectedwrite TDateTime PreviousWorkSubmittedAt;
+
+// Work rate is meaured in hours
+var protectedwrite int CachedWorkRate;
+var protectedwrite int NextSpawnAt; // In work units
+
+var array<ChainDeckEntry> ChainDeck;
+var bool bDarkEventsActive;
+
+var const config array<int> WorkRateXcom;
+var const config array<int> WorkRatePerContact;
+var const config array<int> WorkRatePerRelay;
+var const config bool bStaringRegionContributesToWork;
+
+var const config array<int> GameStartWork; // How much work to add when the campaign starts
+var const config array<int> WorkRequiredForSpawn;
+var const config array<int> WorkRequiredForSpawnVariance;
+
+var config int MinSupplies;
+var config array<name> SupplyChains;
+var config int MinIntel;
+var config array<name> IntelChains;
+
+static function Update()
+{
+	local XComGameState_ActivityChainSpawner Spawner;
+	local XComGameState_HeadquartersResistance ResHQ;
+	local XComGameState_HeadquartersAlien AlienHQ;
+	local XComGameState NewGameState;
+	local UIStrategyMap StrategyMap;
+	local bool bDirty;
+
+	Spawner = GetSpawner(false); // Do not spam redscreens every tick
+	if (Spawner == none)
+	{
+		`RedScreenOnce("CI: Failed to fetch XComGameState_ActivityChainSpawner for ticking");
+		return;
+	}
+
+	NewGameState = class'XComGameStateContext_ChangeContainer'.static.CreateChangeState("CI: XComGameState_ActivityChainSpawner update");
+	Spawner = XComGameState_ActivityChainSpawner(NewGameState.ModifyStateObject(class'XComGameState_ActivityChainSpawner', Spawner.ObjectID));
+	
+	StrategyMap = `HQPRES.StrategyMap2D;
+
+	// STEP 1: we check if we are due spawning an action at CachedWorkRate
+	if (Spawner.ShouldSpawnChain() && StrategyMap != none && StrategyMap.m_eUIState != eSMS_Flight)
+	{
+		`log("Enough work for P1, starting spawning",, 'CI_ACSpawner');
```

Comment:
Update log text, but also no need to use `CI_ACSpawner` - the P1 log category predates CI_Trace/CI_Log and it's better to use the latter instead